internal/gcs-sidecar/bridge.go

```diff
@@ -0,0 +1,439 @@
+	"go.opencensus.io/trace/tracestate"
+	"golang.org/x/sys/windows"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"github.com/Microsoft/hcsshim/internal/bridgeutils/commonutils"
+	"github.com/Microsoft/hcsshim/internal/bridgeutils/gcserr"
+	"github.com/Microsoft/hcsshim/internal/gcs/prot"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/oc"
+	"github.com/Microsoft/hcsshim/pkg/securitypolicy"
+)
+
+type Bridge struct {
+	mu        sync.Mutex
+	hostState *Host
+	// List of handlers for handling different rpc message requests.
+	rpcHandlerList map[prot.RPCProc]HandlerFunc
+
+	// hcsshim and inbox GCS connections respectively.
+	shimConn     io.ReadWriteCloser
+	inboxGCSConn io.ReadWriteCloser
+
+	// Response channels to forward incoming requests to inbox GCS
+	// and send responses back to hcsshim respectively.
+	sendToGCSCh  chan request
+	sendToShimCh chan bridgeResponse
+}
+
+// SequenceID is used to correlate requests and responses.
+type sequenceID uint64
+
+// messageHeader is the common header present in all communications messages.
+type messageHeader struct {
+	Type prot.MsgType
+	Size uint32
+	ID   sequenceID
+}
+
+type bridgeResponse struct {
+	ctx      context.Context
+	header   messageHeader
+	response []byte
+}
+
+type request struct {
+	// Context created once received from the bridge.
+	ctx context.Context
+	// header is the wire format message header that preceded the message for
+	// this request.
+	header messageHeader
+	// activityID is the id of the specific activity for this request.
+	activityID guid.GUID
+	// message is the portion of the request that follows the `Header`.
+	message []byte
+}
+
+func NewBridge(shimConn io.ReadWriteCloser, inboxGCSConn io.ReadWriteCloser, initialEnforcer securitypolicy.SecurityPolicyEnforcer) *Bridge {
+	hostState := NewHost(initialEnforcer)
+	return &Bridge{
+		rpcHandlerList: make(map[prot.RPCProc]HandlerFunc),
+		hostState:      hostState,
+		shimConn:       shimConn,
+		inboxGCSConn:   inboxGCSConn,
+		sendToGCSCh:    make(chan request),
+		sendToShimCh:   make(chan bridgeResponse),
+	}
+}
+
+func NewPolicyEnforcer(initialEnforcer securitypolicy.SecurityPolicyEnforcer) *SecurityPoliyEnforcer {
+	return &SecurityPoliyEnforcer{
+		securityPolicyEnforcerSet: false,
+		securityPolicyEnforcer:    initialEnforcer,
+	}
+}
+
+// UnknownMessage represents the default handler logic for an unmatched request
+// type sent from the bridge.
+func UnknownMessage(r *request) error {
+	log.G(r.ctx).Debugf("bridge: function not supported, header type %v", prot.MsgType(r.header.Type).String())
+	return gcserr.WrapHresult(errors.Errorf("bridge: function not supported, header type: %v", r.header.Type), gcserr.HrNotImpl)
+}
+
+// HandlerFunc is an adapter to use functions as handlers.
+type HandlerFunc func(*request) error
+
+// ServeMsg serves request by calling appropriate handler functions.
+func (b *Bridge) ServeMsg(r *request) error {
+	b.mu.Lock()
```

Comment:
Locking on a mutex during serving the request can lead to a starvation situation. This only works if we are passing along the request to in-box GCS in an async manner.

If we have a situation wherein the request takes time to process and/or the request waits for a long period such as that of `waitOnProcess` command, the next request will never be processed.

Reply:
Addressed in latest push